File: sahayog/sahayog/report/employee_activity_summary/employee_activity_summary.py
import datetime

import frappe
from frappe.utils import getdate, nowdate


# ---------------------------------------------------------
# NORMALIZE EMPLOYEE ID (NT7177 → 7177)
# ---------------------------------------------------------
def normalize_emp_id(emp_id):
    """Return only numeric part from employee ID (e.g., NT7177 → 7177)."""
    return "".join(filter(str.isdigit, emp_id or ""))


def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    chart = get_chart_based_on_filters(data, filters)
    return columns, data, None, chart


# ---------------------------------------------------------
# COLUMNS
# ---------------------------------------------------------
def get_columns():
    return [
        {
            "label": "Employee ID",
            "fieldname": "emp_id",
            "fieldtype": "Data",
            "width": 120,
        },
        {
            "label": "Employee Name",
            "fieldname": "employee_name",
            "fieldtype": "Data",
            "width": 200,
        },
        {
            "label": "Status",
            "fieldname": "status_html",
            "fieldtype": "HTML",
            "width": 120,
        },
        {
            "label": "Sol ID",
            "fieldname": "sol_id",
            "fieldtype": "Data",
            "width": 100,
        },
        {
            "label": "Branch",
            "fieldname": "branch",
            "fieldtype": "Data",
            "width": 120,
        },
        {
            "label": "Zone",
            "fieldname": "zone",
            "fieldtype": "Data",
            "width": 120,
        },
        {
            "label": "Region",
            "fieldname": "region",
            "fieldtype": "Data",
            "width": 120,
        },
        {
            "label": "State",
            "fieldname": "state",
            "fieldtype": "Data",
            "width": 120,
        },
        {
            "label": "District",
            "fieldname": "district",
            "fieldtype": "Data",
            "width": 120,
        },
        # {
        #     "label": "Joining Date",
        #     "fieldname": "date_of_joining",
        #     "fieldtype": "Date",
        #     "width": 120,
        # },
        # {
        #     "label": "Employee Creation Date",
        #     "fieldname": "creation_date",
        #     "fieldtype": "Date",
        #     "width": 140,
        # },
        # {
        #     "label": "Days Since Joining",
        #     "fieldname": "days_since_joining",
        #     "fieldtype": "Int",
        #     "width": 130,
        # },
        {
            "label": "Total Leads",
            "fieldname": "lead_count",
            "fieldtype": "Int",
            "width": 100,
        },
    ]


# ---------------------------------------------------------
# DATA LOGIC
# ---------------------------------------------------------
def get_data(filters=None):
    # Convert date filter to proper date object
    if filters:
        if filters.get("selected_date") and isinstance(filters["selected_date"], str):
            filters["selected_date"] = datetime.datetime.strptime(
                filters["selected_date"], "%Y-%m-%d"
            ).date()

    today = getdate(nowdate())

    # Prevent future dates
    if filters:
        if filters.get("selected_date") and filters["selected_date"] > today:
            frappe.throw("Selected Date cannot be in the future.")

    # Fetch leads
    leads = frappe.db.get_all(
        "Lead", fields=["name", "lead_name", "status", "lead_owner"], order_by="name"
    )

    # Fetch employees including creation date
    employees = frappe.db.get_all(
        "Employee",
        fields=[
            "name as emp_id",
            "employee_name",
            "sol_id",
            "date_of_joining",
            "creation as creation_date",
            "department",
        ],
        order_by="sol_id asc",
    )

    all_employees = employees[:]

    # Only Sales + Operations employees for non-active
    sales_ops_employees = [
        emp for emp in employees if emp.get("department") in ("Sales", "Operations")
    ]

    sahayog_branches = frappe.db.get_all(
        "Sahayog Branch",
        fields=["sol_id", "zone", "region", "state", "district", "branch"],
    )

    branch_map = {str(b["sol_id"]).strip(): b for b in sahayog_branches}
    emp_map = {normalize_emp_id(emp["emp_id"]): emp for emp in all_employees}

    # Group leads by employee
    emp_leads = {}
    for row in leads:
        nid = normalize_emp_id(row["lead_owner"])
        if not nid:
            continue
        emp_leads.setdefault(nid, {"leads": []})["leads"].append(row)

    report_rows = []
    active_emp_ids = set()

    # Non active lookup
    non_active_emp_info = {}
    for emp in all_employees:
        nid = normalize_emp_id(emp["emp_id"])
        if nid not in emp_leads:
            non_active_emp_info[nid] = emp

    # ------------------------------
    # ACTIVE EMPLOYEES
    # ------------------------------
    for emp_id in emp_leads:
        numeric_id = normalize_emp_id(emp_id)
        emp_ref = emp_map.get(numeric_id)

        if emp_ref:
            employee_name = emp_ref["employee_name"]
            date_of_joining = emp_ref["date_of_joining"]
            sol_id = str(emp_ref.get("sol_id")).strip()
            creation_date = emp_ref.get("creation_date")
        else:
            employee_name = "Unknown"
            date_of_joining = None
            sol_id = None
            creation_date = None

        branch_data = branch_map.get(sol_id, {})
        days_since_joining = (today - date_of_joining).days if date_of_joining else 0

        report_rows.append(
            {
                "emp_id": numeric_id,
                "employee_name": employee_name,
                "sol_id": sol_id,
                "zone": branch_data.get("zone", ""),
                "region": branch_data.get("region", ""),
                "state": branch_data.get("state", ""),
                "district": branch_data.get("district", ""),
                "branch": branch_data.get("branch", ""),
                "date_of_joining": date_of_joining,
                "creation_date": creation_date,
                "days_since_joining": days_since_joining,
                "lead_count": len(emp_leads[numeric_id]["leads"]),
                "status": "Active",
                "status_html": (
                    "<div style='display:inline-flex; align-items:center; gap:6px; "
                    "padding:0px 14px; background:#4caf50; border:2px solid #4caf50; "
                    "color:#ffffff; font-weight:400; border-radius:10px;'>ACTIVE</div>"
                ),
            }
        )

        active_emp_ids.add(numeric_id)

    # ------------------------------
    # NON-ACTIVE EMPLOYEES
    # ------------------------------
    for emp in sales_ops_employees:
        nid = normalize_emp_id(emp["emp_id"])

        if nid not in active_emp_ids:
            sol_id = str(emp.get("sol_id")).strip()
            days_since_joining = (
                (today - emp["date_of_joining"]).days if emp["date_of_joining"] else 0
            )
            branch_data = branch_map.get(sol_id, {})

            report_rows.append(
                {
                    "emp_id": nid,
                    "employee_name": emp["employee_name"],
                    "sol_id": sol_id,
                    "zone": branch_data.get("zone", ""),
                    "region": branch_data.get("region", ""),
                    "state": branch_data.get("state", ""),
                    "district": branch_data.get("district", ""),
                    "branch": branch_data.get("branch", ""),
                    "date_of_joining": emp["date_of_joining"],
                    "creation_date": emp.get("creation_date"),
                    "days_since_joining": days_since_joining,
                    "lead_count": 0,
                    "status": "Non-Active",
                    "status_html": (
                        "<div style='display:inline-flex; align-items:center; gap:6px; "
                        "padding:0px 14px; background:#F18F01; border:2px solid #F18F01; "
                        "color:#ffffff; font-weight:400; border-radius:10px;'>INACTIVE</div>"
                    ),
                }
            )

    # FILTERING
    if filters:
        filtered = []
        for row in report_rows:
            show = True

            if filters.get("sol_id") and str(row.get("sol_id")) != str(filters["sol_id"]):
                show = False

            if filters.get("zone") and row.get("zone") != filters["zone"]:
                show = False

            if filters.get("region") and row.get("region") != filters["region"]:
                show = False

            if filters.get("branch") and row.get("branch") != filters["branch"]:
                show = False

            if filters.get("status") and row.get("status") != filters["status"]:
                show = False

            # SINGLE DATE FILTER - Show only records created on the selected date
            if filters.get("selected_date") and row.get("creation_date"):
                if row.get("creation_date").date() != filters.get("selected_date"):
                    show = False

            if show:
                filtered.append(row)

        return filtered

    return report_rows


# ---------------------------------------------------------
# CHART LOGIC
# ---------------------------------------------------------
def get_chart_based_on_filters(data, filters=None):
    def grouped_chart(group_field, label_name):
        group_status_users = {}

        for row in data:
            group_val = row.get(group_field, f"Unknown {label_name}")
            status = row.get("status", "Unknown")
            emp_id = row.get("emp_id")

            if not emp_id:
                continue

            group_status_users.setdefault(
                group_val, {"Active": set(), "Non-Active": set()}
            )

            group_status_users[group_val][status].add(emp_id)

        valid_groups = [
            g for g in group_status_users.keys()
            if str(g).strip() and not str(g).lower().startswith("unknown")
        ]

        labels = sorted(valid_groups)
        active_counts = [len(group_status_users[g]["Active"]) for g in labels]
        non_active_counts = [len(group_status_users[g]["Non-Active"]) for g in labels]

        return {
            "data": {
                "labels": labels,
                "datasets": [
                    {"name": "Active Users", "values": active_counts},
                    {"name": "Non-Active Users", "values": non_active_counts},
                ],
            },
            "type": "bar",
            "colors": ["#62B58F", "#F18F01"],
        }

    if filters:
        if filters.get("branch"):
            return None
        if filters.get("sol_id"):
            return None
        if filters.get("zone") and not filters.get("region"):
            return None
        if filters.get("district"):
            return grouped_chart("branch", "Branch")
        if filters.get("region"):
            return grouped_chart("branch", "Branch")

    return grouped_chart("zone", "Zone")

File: sahayog/sahayog/report/employee_activity_summary/employee_activity_summary.js
// Copyright (c) 2025, Developer Team and contributors
// For license information, please see license.txt

frappe.query_reports["Employee Activity Summary"] = {
  filters: [
    // {
    //   fieldname: "from_date",
    //   label: __("From Date"),
    //   fieldtype: "Date",
    //   default: frappe.datetime.get_today(),
    // },
    // {
    //   fieldname: "to_date",
    //   label: __("To Date"),
    //   fieldtype: "Date",
    //   default: frappe.datetime.get_today(),
    // },
    {
      fieldname: "selected_date", // ← CHANGED from "from_date"
      label: __("Select Date"), // ← CHANGED label
      fieldtype: "Date",
      default: frappe.datetime.get_today(),
    },
    {
      fieldname: "zone",
      label: __("Zone"),
      fieldtype: "Autocomplete",
      options: [""],
    },
    {
      fieldname: "region",
      label: __("Region"),
      fieldtype: "Autocomplete",
      options: [""],
    },
    // SEARCHABLE AUTOCOMPLETE FIELDS
    {
      fieldname: "sol_id",
      label: __("Sol ID"),
      fieldtype: "Autocomplete",
      options: [""],
    },
    {
      fieldname: "branch",
      label: __("Branch"),
      fieldtype: "Autocomplete",
      options: [""],
    },

    {
      fieldname: "status",
      label: __("Status"),
      fieldtype: "Select",
      options: ["", "Active", "Non-Active"],
    },
  ],

  onload(report) {
    // HIDE REGION & BRANCH INITIALLY
    frappe.query_report.get_filter("region").toggle(false);
    frappe.query_report.get_filter("branch").toggle(false);

    frappe.call({
      method: "frappe.client.get_list",
      args: {
        doctype: "Sahayog Branch",
        fields: ["sol_id", "zone", "region", "branch"],
        limit_page_length: 2000,
      },

      callback(r) {
        if (!r.message) return;

        let sol_ids = new Set();
        let zones = new Set();
        let regions = new Set();
        let branches = new Set();

        let zone_region_map = {};
        let region_branch_map = {};
        let sol_map = {};

        r.message.forEach((row) => {
          if (row.sol_id) sol_ids.add(row.sol_id);
          if (row.zone) zones.add(row.zone);
          if (row.region) regions.add(row.region);
          if (row.branch) branches.add(row.branch);

          if (row.zone && row.region) {
            if (!zone_region_map[row.zone])
              zone_region_map[row.zone] = new Set();
            zone_region_map[row.zone].add(row.region);
          }

          if (row.region && row.branch) {
            if (!region_branch_map[row.region])
              region_branch_map[row.region] = new Set();
            region_branch_map[row.region].add(row.branch);
          }

          sol_map[row.sol_id] = {
            zone: row.zone,
            region: row.region,
            branch: row.branch,
          };
        });

        // Works with Autocomplete
        function update_select(fieldname, values) {
          let field = frappe.query_report.get_filter(fieldname);
          if (!field) return;

          field.set_data(["", ...values]);
        }

        update_select("sol_id", Array.from(sol_ids).sort());
        update_select("zone", Array.from(zones).sort());
        update_select("region", Array.from(regions).sort());
        update_select("branch", Array.from(branches).sort());

        frappe.query_report.get_filter("region").$input.prop("disabled", true);
        frappe.query_report.get_filter("branch").$input.prop("disabled", true);

        // ====================================================
        // ZONE → FILTER REGIONS & SHOW REGION
        // ====================================================
        frappe.query_report.get_filter("zone").$input.on("change", function () {
          let zone_val = frappe.query_report.get_filter_value("zone");

          frappe.query_report.set_filter_value("region", "");
          frappe.query_report.set_filter_value("branch", "");

          if (zone_val) {
            let filtered_regions = Array.from(
              zone_region_map[zone_val] || []
            ).sort();

            update_select("region", filtered_regions);

            frappe.query_report
              .get_filter("region")
              .$input.prop("disabled", false);
            frappe.query_report
              .get_filter("branch")
              .$input.prop("disabled", true);

            // SHOW region
            frappe.query_report.get_filter("region").toggle(true);
            frappe.query_report.get_filter("branch").toggle(false);
          } else {
            frappe.query_report
              .get_filter("region")
              .$input.prop("disabled", true);
            frappe.query_report
              .get_filter("branch")
              .$input.prop("disabled", true);

            // HIDE region & branch
            frappe.query_report.get_filter("region").toggle(false);
            frappe.query_report.get_filter("branch").toggle(false);
          }
        });

        // ====================================================
        // REGION → FILTER BRANCHES & SHOW BRANCH
        // ====================================================
        frappe.query_report
          .get_filter("region")
          .$input.on("change", function () {
            let region_val = frappe.query_report.get_filter_value("region");

            frappe.query_report.set_filter_value("branch", "");

            if (region_val) {
              let filtered_branches = Array.from(
                region_branch_map[region_val] || []
              ).sort();

              update_select("branch", filtered_branches);
              frappe.query_report
                .get_filter("branch")
                .$input.prop("disabled", false);

              // SHOW branch
              frappe.query_report.get_filter("branch").toggle(true);
            } else {
              frappe.query_report
                .get_filter("branch")
                .$input.prop("disabled", true);

              // HIDE branch
              frappe.query_report.get_filter("branch").toggle(false);
            }
          });

        // ====================================================
        // SOL → AUTO-FILL
        // ====================================================
        frappe.query_report
          .get_filter("sol_id")
          .$input.on("change", function () {
            let sol_val = frappe.query_report.get_filter_value("sol_id");

            if (!sol_val || !sol_map[sol_val]) {
              frappe.query_report.set_filter_value("zone", "");
              frappe.query_report.set_filter_value("region", "");
              frappe.query_report.set_filter_value("branch", "");

              frappe.query_report.get_filter("region").toggle(false);
              frappe.query_report.get_filter("branch").toggle(false);
              return;
            }

            let { zone, region, branch } = sol_map[sol_val];

            frappe.query_report.set_filter_value("zone", zone);

            let filtered_regions = Array.from(
              zone_region_map[zone] || []
            ).sort();
            update_select("region", filtered_regions);
            frappe.query_report
              .get_filter("region")
              .$input.prop("disabled", false);
            frappe.query_report.set_filter_value("region", region);

            let filtered_branches = Array.from(
              region_branch_map[region] || []
            ).sort();
            update_select("branch", filtered_branches);
            frappe.query_report
              .get_filter("branch")
              .$input.prop("disabled", false);
            frappe.query_report.set_filter_value("branch", branch);

            // SHOW both
            frappe.query_report.get_filter("region").toggle(true);
            frappe.query_report.get_filter("branch").toggle(true);
          });
      },
    });

    // ===================================================================
    // CARD & CHART LOGIC (UNCHANGED)
    // ===================================================================

    function render_emp_status_cards(
      active_count,
      non_active_count,
      branch_name
    ) {
      $("#emp-status-card-container").remove();

      var card_html = `
        <div id="emp-status-card-container" style="margin-bottom:12px; background:#fff; padding:12px; border-radius:6px; border:1px solid #e2e2e2;">
          <div style="display:flex; align-items:center; gap:20px;">
            <div style="display:flex; align-items:center; gap:6px;">
              <div style="width:12px; height:12px; background:#62B58F; border-radius:2px;"></div>
              <span style="font-size:13px; color:#666;">Active:</span>
              <span style="font-size:14px; color:#62B58F; font-weight:600;">${active_count}</span>
            </div>
            <div style="height:20px; width:1px; background:#e2e2e2;"></div>
            <div style="display:flex; align-items:center; gap:6px;">
              <div style="width:12px; height:12px; background:#F18F01; border-radius:2px;"></div>
              <span style="font-size:13px; color:#666;">Non-Active:</span>
              <span style="font-size:14px; color:#F18F01; font-weight:600;">${non_active_count}</span>
            </div>
            <div style="display:flex; align-items:center; gap:6px; margin-left:auto;">
              <span style="font-size:13px; color:#666;">Total:</span>
              <span style="font-size:14px; color:#333; font-weight:600;">
                ${active_count + non_active_count}
              </span>
            </div>
          </div>
        </div>
      `;

      let $target = $(".frappe-query-report");
      if ($target.length === 0) $target = $(".page-content .report-wrapper");
      if ($target.length === 0) $target = $(".layout-main-section");

      if ($target.length > 0) $target.prepend(card_html);
    }

    function render_region_chart(data, zone_name) {
      $("#region-chart-container").remove();

      let region_status_users = {};

      data.forEach((row) => {
        let region = row.region || "Unknown Region";
        let status = row.status || "Unknown";
        let emp_id = row.emp_id;

        if (!emp_id) return;

        if (!region_status_users[region]) {
          region_status_users[region] = {
            Active: new Set(),
            "Non-Active": new Set(),
          };
        }

        if (status === "Active")
          region_status_users[region]["Active"].add(emp_id);
        else if (status === "Non-Active")
          region_status_users[region]["Non-Active"].add(emp_id);
      });

      let regions = Object.keys(region_status_users).sort();
      let active_counts = regions.map(
        (r) => region_status_users[r]["Active"].size
      );
      let non_active_counts = regions.map(
        (r) => region_status_users[r]["Non-Active"].size
      );

      let chart_html = `
        <div id="region-chart-container" style="margin-bottom:15px; background:#fff; padding:20px; border-radius:8px; border:1px solid #d1d8dd;">
          <h3 style="margin:0 0 15px 0; color:#333; font-size:16px; font-weight:600;">Region-wise Employee Distribution - ${zone_name}</h3>
          <div id="region-chart" style="height:300px;"></div>
        </div>
      `;

      let $target = $(".frappe-query-report");
      if ($target.length === 0) $target = $(".page-content .report-wrapper");
      if ($target.length === 0) $target = $(".layout-main-section");

      if ($target.length > 0) {
        $target.prepend(chart_html);

        setTimeout(() => {
          new frappe.Chart("#region-chart", {
            data: {
              labels: regions,
              datasets: [
                {
                  name: "Active Users",
                  values: active_counts,
                  chartType: "bar",
                },
                {
                  name: "Non-Active Users",
                  values: non_active_counts,
                  chartType: "bar",
                },
              ],
            },
            type: "bar",
            height: 280,
            colors: ["#62B58F", "#F18F01"],
          });
        }, 100);
      }
    }

    let original_refresh = report.refresh;

    report.refresh = function () {
      original_refresh.call(report);

      let zone = report.get_filter_value("zone");
      let region = report.get_filter_value("region");
      let branch = report.get_filter_value("branch");

      if (branch) {
        setTimeout(() => {
          let data = report.data || [];
          let active = data.filter((x) => x.status === "Active").length;
          let non_active = data.filter((x) => x.status === "Non-Active").length;
          render_emp_status_cards(active, non_active, branch);
        }, 400);
      } else if (zone && !region) {
        setTimeout(() => {
          let data = report.data || [];
          render_region_chart(data, zone);
        }, 400);

        $("#emp-status-card-container").remove();
      } else {
        $("#emp-status-card-container").remove();
        $("#region-chart-container").remove();
      }
    };
  },
};
